zookeeper/src/main/java/com/linecorp/armeria/server/zookeeper/ZooKeeperUpdatingListenerBuilder.java

```diff
@@ -0,0 +1,206 @@
+public final class ZooKeeperUpdatingListenerBuilder {
+    private CuratorFramework client;
+    private String connectionStr;
+    private String zNodePath;
+    private int connectTimeoutMillis = ZooKeeperDefaults.DEFAULT_CONNECT_TIMEOUT_MS;
+    private int sessionTimeoutMillis = ZooKeeperDefaults.DEFAULT_SESSION_TIMEOUT_MS;
+    private Endpoint endpoint;
+    private NodeValueCodec nodeValueCodec = NodeValueCodec.DEFAULT;
+
+    /**
+     * Creates a {@link ZooKeeperUpdatingListenerBuilder} with a {@link CuratorFramework} instance and a zNode
+     * path.
+     *
+     * @param client the curator framework instance
+     * @param zNodePath the ZooKeeper node to register
+     */
+    public ZooKeeperUpdatingListenerBuilder(CuratorFramework client, String zNodePath) {
+        this.client = requireNonNull(client, "client");
+        this.zNodePath = requireNonNull(zNodePath, "zNodePath");
+        checkArgument(!this.zNodePath.isEmpty(), "zNodePath can't be empty");
+    }
+
+    /**
+     * Creates a {@link ZooKeeperUpdatingListenerBuilder} with a ZooKeeper connection string and a zNode path.
+     *
+     * @param connectionStr the ZooKeeper connection string
+     * @param zNodePath the ZooKeeper node to register
+     */
+    public ZooKeeperUpdatingListenerBuilder(String connectionStr, String zNodePath) {
+        this.connectionStr = requireNonNull(connectionStr, "connectionStr");
+        checkArgument(!this.connectionStr.isEmpty(), "connectionStr can't be empty");
+        this.zNodePath = requireNonNull(zNodePath, "zNodePath");
+        checkArgument(!this.zNodePath.isEmpty(), "zNodePath can't be empty");
+    }
+
+    /**
+     * Sets the connect timeout.
+     *
+     * @param connectTimeout the connect timeout
+     *
+     * @throws IllegalStateException if this builder is constructed with
+     *                               {@link #ZooKeeperUpdatingListenerBuilder(CuratorFramework, String)}
+     */
+    public ZooKeeperUpdatingListenerBuilder connectTimeout(Duration connectTimeout) {
+        requireNonNull(connectTimeout, "connectTimeout");
+        checkArgument(!connectTimeout.isZero() && !connectTimeout.isNegative(),
+                      "connectTimeout: %s (expected: > 0)", connectTimeout);
+        return connectTimeoutMillis(connectTimeout.toMillis());
+    }
+
+    /**
+     * Sets the connect timeout (in ms). (default: 1000)
+     *
+     * @param connectTimeoutMillis the connect timeout
+     *
+     * @throws IllegalStateException if this builder is constructed with
+     *                               {@link #ZooKeeperUpdatingListenerBuilder(CuratorFramework, String)}
+     */
+    public ZooKeeperUpdatingListenerBuilder connectTimeoutMillis(long connectTimeoutMillis) {
+        checkArgument(connectTimeoutMillis > 0,
+                      "connectTimeoutMillis: %s (expected: > 0)", connectTimeoutMillis);
+        if (client != null) {
+            throw new IllegalStateException("client is already set");
```

Comment:
- Could use `Preconditions.checkState()`.
- `This method is allowed only when created with a connection string.`
- Could be extracted to a common method like `private void ensureInternalClient()`

Reply:
You mean, similar to `Endpoint#ensureGroup` or `Endpoint#ensureSingle`. Right?